lib/golem/doc/system-design.md

```diff
@@ -0,0 +1,80 @@
+# Arm control system design
+
+We want to achieve the following with our arm control system:
+
+- Track the state (position, velocity, torque) of the arm
+- Move the arm to desired positions while satisfying constraints:
+    * Avoid obstacles
+    * Limit maximum velocity and torque on the joints
+- Detect when the arm is blocked
+
+For that, we'll get inspiration from [Mikrob technology's control system design in aversive](https://wiki.droids-corp.org/articles/a/v/e/Aversive/Asservissement_Microb_2008.html).
+It provides us with a modular clean architecture that separates concerns properly.
+The separate components can be composed to implement complex behavior, and some components are optional.
+Here follows a diagram of the proposed design.
+
+![Diagram of the proposed arm control system design](figures/system.pdf)
+
+Each of the main blocks (motors, control loop, trajectory, and blocking detection) run in their own thread.
```

Comment:
How do they communicate ?

Reply:
Magic! Through function calls. For example you can set the controller's target via a function call from another thread. Still need to think about synchronization in the implementation :thinking: